sled-agent/src/instance.rs

```diff
@@ -492,32 +525,147 @@ impl Instance {
             state: InstanceStates::new(initial.runtime),
             running_state: None,
             lazy_nexus_client,
+            instance_ticket: ticket,
         };
 
         let inner = Arc::new(Mutex::new(instance));
 
         Ok(Instance { inner })
     }
 
+    pub async fn current_state(&self) -> InstanceRuntimeState {
+        let inner = self.inner.lock().await;
+        inner.state.current().clone()
+    }
+
+    /// Ensures that a Propolis process exists for this instance, then sends it
+    /// an instance ensure request.
+    async fn propolis_ensure(
+        &self,
+        inner: &mut MutexGuard<'_, InstanceInner>,
+        migration_params: Option<InstanceMigrationTargetParams>,
+    ) -> Result<(), Error> {
+        if let Some(running_state) = inner.running_state.as_ref() {
+            inner
+                .propolis_ensure(
+                    &running_state.client,
+                    &running_state.running_zone,
+                    migration_params,
+                )
+                .await?;
+        } else {
+            // If there's no Propolis yet, and this instance is not being
+            // initialized via migration, immediately send a state update to
+            // Nexus to reflect that the instance is starting (so that the
+            // external API will display this state while the zone is being
+            // started).
+            //
+            // Migration targets don't do this because the instance is still
+            // logically running (on the source) while the target Propolis is
+            // being launched.
+            if migration_params.is_none() {
+                inner.state.transition(InstanceState::Starting);
+                inner.publish_state_to_nexus().await?;
+            }
+
+            let setup_result: Result<(), Error> = {
```

Comment:
Is this supposed to be wrapped up in a lambda?

I believe this code is equivalent to the following: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5e000e905c95ffe9541c0d62f66f834a

In which the errors from setting up propolis will never be assigned to `setup_result`, but which will instead cause the `propolis_ensure` function to exit.

In contrast, the following actually does let the result get a non-`Ok(())` value:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=50f83c62b40e5ebf27a22876828b0542

Reply:
You're right, this is incorrect. Unless I'm missing something (very possible here!), turning it into a closure is nontrivial because [async closures are still unstable](https://github.com/rust-lang/rust/issues/62290). From what I've read, it might be possible to use a closure anyway if I write out a sufficiently gnarly `Future` return type for it, but in 7672fcf I went with a label-based approach that I think will do the job.